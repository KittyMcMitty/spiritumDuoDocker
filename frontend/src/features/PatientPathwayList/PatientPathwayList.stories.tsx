/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { NewMockSdApolloProvider } from 'test/mocks/mockApolloProvider';
import { DocumentNode } from 'graphql';
import { RequestHandler } from 'mock-apollo-client';
import { ClinicalRequestState } from '__generated__/globalTypes';
import { BrowserRouter } from 'react-router-dom';

import {
  getPatientOnPathwayConnectionForGrp,
  getPatientOnPathwayConnectionForGrpVariables,
  getPatientOnPathwayConnectionForGrp_getPatientOnPathwayConnection_edges_node as PatientNode,
  getPatientOnPathwayConnectionForGrp_getPatientOnPathwayConnection_edges as PatientEdge,
} from './__generated__/getPatientOnPathwayConnectionForGrp';
import PatientPathwayList, { GET_PATIENT_ON_PATHWAY_CONNECTION_QUERY } from './PatientPathwayList';

const totalCount = 12;

const mockPatient: (id: string) => PatientNode = (id) => ({
  __typename: 'Patient',
  id: id,
  firstName: 'John',
  lastName: `Doe ${id}`,
  dateOfBirth: new Date('1970-10-31'),
  hospitalNumber: 'fMRN1234567',
  nationalNumber: 'fNHS1234567890',
  onPathways: [{
    __typename: 'OnPathway',
    id: '1',
    referredAt: new Date('2021-08-14'),
    clinicalRequests: [
      {
        __typename: 'ClinicalRequest',
        id: '1',
        clinicalRequestType: { __typename: 'ClinicalRequestType', id: '1', name: 'CT-Thorax' },
        addedAt: new Date('2021-08-14'),
        updatedAt: new Date('2021-08-16'),
        currentState: ClinicalRequestState.COMPLETED,
        forwardDecisionPoint: {
          __typename: 'DecisionPoint',
          id: '1',
          addedAt: new Date('2021-08-24'),
        },
      },
      {
        __typename: 'ClinicalRequest',
        id: '2',
        clinicalRequestType: { __typename: 'ClinicalRequestType', id: '2', name: 'X-Ray' },
        addedAt: new Date('2021-08-14'),
        updatedAt: new Date('2021-08-20'),
        currentState: ClinicalRequestState.COMPLETED,
        forwardDecisionPoint: {
          __typename: 'DecisionPoint',
          id: '1',
          addedAt: new Date('2021-08-24'),
        },
      },
      {
        __typename: 'ClinicalRequest',
        id: '3',
        clinicalRequestType: { __typename: 'ClinicalRequestType', id: '3', name: 'Lung Function' },
        addedAt: new Date('2021-08-14'),
        updatedAt: new Date('2021-08-15'),
        currentState: ClinicalRequestState.COMPLETED,
        forwardDecisionPoint: {
          __typename: 'DecisionPoint',
          id: '1',
          addedAt: new Date('2021-08-24'),
        },
      },
      {
        __typename: 'ClinicalRequest',
        id: '4',
        clinicalRequestType: { __typename: 'ClinicalRequestType', id: '4', name: 'CT-Thorax' },
        addedAt: new Date('2021-08-24'),
        updatedAt: new Date('2021-08-30'),
        currentState: ClinicalRequestState.COMPLETED,
        forwardDecisionPoint: {
          __typename: 'DecisionPoint',
          id: '2',
          addedAt: new Date('2021-09-06'),
        },
      },
      {
        __typename: 'ClinicalRequest',
        id: '5',
        clinicalRequestType: { __typename: 'ClinicalRequestType', id: '5', name: 'MRI-Head' },
        addedAt: new Date('2021-08-24'),
        updatedAt: new Date('2021-09-02'),
        currentState: ClinicalRequestState.COMPLETED,
        forwardDecisionPoint: {
          __typename: 'DecisionPoint',
          id: '2',
          addedAt: new Date('2021-09-06'),
        },
      },
      {
        __typename: 'ClinicalRequest',
        id: '6',
        clinicalRequestType: { __typename: 'ClinicalRequestType', id: '6', name: 'EBUS' },
        addedAt: new Date('2021-08-24'),
        updatedAt: new Date('2021-09-04'),
        currentState: ClinicalRequestState.COMPLETED,
        forwardDecisionPoint: {
          __typename: 'DecisionPoint',
          id: '2',
          addedAt: new Date('2021-09-06'),
        },
      },
      {
        __typename: 'ClinicalRequest',
        id: '7',
        clinicalRequestType: { __typename: 'ClinicalRequestType', id: '7', name: 'CT-Thorax' },
        addedAt: new Date('2021-09-10'),
        updatedAt: new Date('2021-09-10'),
        currentState: ClinicalRequestState.WAITING,
        forwardDecisionPoint: null,
      },
      {
        __typename: 'ClinicalRequest',
        id: '8',
        clinicalRequestType: { __typename: 'ClinicalRequestType', id: '8', name: 'Bloods' },
        addedAt: new Date('2021-09-10'),
        updatedAt: new Date('2021-09-10'),
        currentState: ClinicalRequestState.WAITING,
        forwardDecisionPoint: null,
      },
    ],
  }],
});

// eslint-disable-next-line max-len, camelcase
const mockPatientEdges: PatientEdge[] = [];
for (let i = 1; i <= totalCount; i += 1) {
  // eslint-disable-next-line camelcase
  const edge: PatientEdge = {
    __typename: 'PatientEdge',
    cursor: i.toString(),
    node: mockPatient(i.toString()),
  };
  mockPatientEdges.push(edge);
}

const apolloMock: {
  query: DocumentNode,
  mockFn: RequestHandler<
    getPatientOnPathwayConnectionForGrp,
    getPatientOnPathwayConnectionForGrpVariables
  >
}[] = [{
  query: GET_PATIENT_ON_PATHWAY_CONNECTION_QUERY,
  mockFn: ({ first, after }) => new Promise((resolve) => {
    const start = after
      ? parseInt(after, 10)
      : 0;
    const end = start + first;
    resolve({ data: {
      getPatientOnPathwayConnection: {
        __typename: 'PatientConnection',
        totalCount: totalCount,
        pageInfo: {
          __typename: 'PageInfo',
          hasNextPage: end < totalCount,
          endCursor: end.toString(),
        },
        edges: mockPatientEdges.slice(start, end),
      },
    } });
  }),
}];

export default {
  title: 'Features/PatientPathwayList',
  component: PatientPathwayList,
  decorators: [(Story) => (
    <NewMockSdApolloProvider mocks={ apolloMock }>
      <BrowserRouter>
        <div style={ { width: '90vw', height: '100vh', minWidth: '300px' } }>
          <Story />
        </div>
      </BrowserRouter>
    </NewMockSdApolloProvider>
  )],
} as ComponentMeta<typeof PatientPathwayList>;

const Template: ComponentStory<typeof PatientPathwayList> = (args) => (
  <PatientPathwayList { ...args } />
);

export const Default = Template.bind({});
Default.args = {
  outstanding: true,
  pathwayId: '1',
  patientsToDisplay: 5,
  underCareOf: false,
  includeDischarged: false,
};
