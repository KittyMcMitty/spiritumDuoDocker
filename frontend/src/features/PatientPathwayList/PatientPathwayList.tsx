import React, { useMemo, useState } from 'react';
import { gql, useQuery } from '@apollo/client';
import edgesToNodes from 'app/pagination';

import PathwayVisualisation, { PATIENT_PARTS_FOR_GRP, patientToVisData, PatientVisData } from 'components/PathwayVisualisation';
import { ParentSize } from '@visx/responsive';
import ReactPaginate from 'react-paginate';

import { useAppDispatch } from 'app/hooks';
import { setModalPatientHospitalNumber } from 'pages/HomePage.slice';
import { getPatientOnPathwayConnectionForGrp } from './__generated__/getPatientOnPathwayConnectionForGrp';

export const GET_PATIENT_ON_PATHWAY_CONNECTION_QUERY = gql`
  ${PATIENT_PARTS_FOR_GRP}
  query getPatientOnPathwayConnectionForGrp(
    $outstanding: Boolean, $pathwayId:ID!, $first:Int!, $after: String, $underCareOf: Boolean, $includeDischarged: Boolean
  ) {
    getPatientOnPathwayConnection(
      outstanding: $outstanding, pathwayId:$pathwayId, first:$first, after: $after, underCareOf: $underCareOf, includeDischarged: $includeDischarged
    ) @connection(
        key: "getPatientOnPathwayConnectionForGrp",
        filter: ["outstanding", "pathwayId", "underCareOf", "includeDischarged"]
        )
    {
      totalCount
      pageInfo {
        hasNextPage
        endCursor
      }
      edges {
        cursor
        node {
          ...GrpPatientFields
        }
      }
    }
  }
`;

interface PatientPathwayListProps {
  outstanding?: boolean;
  pathwayId: string;
  patientsToDisplay: number;
  underCareOf?: boolean;
  includeDischarged?: boolean;
}

const PatientPathwayList = ({
  outstanding, pathwayId, patientsToDisplay, underCareOf, includeDischarged,
}: PatientPathwayListProps) => {
  const [currentPage, setCurrentPage] = useState(0);
  const [maxFetchedPage, setMaxFetchedPage] = useState(0);
  const { loading, error, data, fetchMore } = useQuery<getPatientOnPathwayConnectionForGrp>(
    GET_PATIENT_ON_PATHWAY_CONNECTION_QUERY,
    { variables: {
      pathwayId: pathwayId,
      outstanding: outstanding,
      first: patientsToDisplay,
      underCareOf: underCareOf,
      includeDischarged: includeDischarged,
    } },
  );

  const dispatch = useAppDispatch();

  type PatientNode = getPatientOnPathwayConnectionForGrp['getPatientOnPathwayConnection']['edges'][0]['node'];
  const { nodes, pageCount, pageInfo } = edgesToNodes<PatientNode>(
    data?.getPatientOnPathwayConnection, currentPage, patientsToDisplay,
  );

  const patientData = useMemo(() => nodes?.map((n) => patientToVisData(n)), [nodes]);
  const rowClicked = (patient: PatientVisData): void => {
    dispatch(setModalPatientHospitalNumber(patient.hospitalNumber));
  };
  if (patientData) {
    return (
      <div>
        <ParentSize>
          {({ width: parentWidth }) => (
            <PathwayVisualisation
              data={ patientData }
              maxDays={ 80 }
              onClick={ rowClicked }
              showName
              // showNationalNumber
              width={ parentWidth }
              margin={ { top: 40, right: 15, bottom: 30, left: 15 } }
            />
          )}
        </ParentSize>
        <ReactPaginate
          previousLabel="previous"
          nextLabel="next"
          breakLabel="..."
          breakClassName="break-me"
          pageCount={ Math.ceil(pageCount) }
          marginPagesDisplayed={ 2 }
          pageRangeDisplayed={ 5 }
          onPageChange={ ({ selected }) => {
            setCurrentPage(selected);
            if (selected > maxFetchedPage) {
              if (pageInfo?.hasNextPage) {
                setMaxFetchedPage(selected);
                fetchMore({
                  variables: {
                    after: pageInfo?.endCursor,
                  },
                });
              }
            }
          } }
          containerClassName="pagination justify-content-center"
          activeClassName="bkcolour-nhs-blue"
          pageClassName="page-item"
          previousClassName="page-item"
          nextClassName="page-item"
          previousLinkClassName="page-link colour-nhs-blue paginator-button"
          nextLinkClassName="page-link colour-nhs-blue paginator-button"
          pageLinkClassName="page-link colour-nhs-blue paginator-button"
          activeLinkClassName="bkcolour-nhs-blue text-white paginator-button"
        />
      </div>
    );
  }
  return <div>Loading</div>;
};

export default PatientPathwayList;
