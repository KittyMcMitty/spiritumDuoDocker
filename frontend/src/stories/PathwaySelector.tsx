import React from 'react';
import { Link } from "react-router-dom";

interface PathwaySelectorProps {
	/**
	* List of pathways to present the user
	*/
	options: Array <string>;
	/**
	* Current selected option
	*/
	currentOption: string;
	/**
	* Function to handle item select
	*/
	onItemSelect(pathwayName:string): void;
}

/**
* Primary UI component for user interaction
*/
export const PathwaySelector = (
	{
		options = ["Default"],
		currentOption = "Default"
}: PathwaySelectorProps) => {
	let itemList=options.map((item,index)=>{
		return <Link className="dropdown-item" to={"/page/"}>Refer</Link>;
	})
	
	return(
		<div className="dropdown">
			<button className="btn btn-outline-secondary dropdown-toggle" id="dropdownMenuButton" type="button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
				{currentOption}
			</button>
			<div className="dropdown-menu" aria-labelledby="dropdownMenuButton">
				{itemList}
			</div>
		</div>
	);
};