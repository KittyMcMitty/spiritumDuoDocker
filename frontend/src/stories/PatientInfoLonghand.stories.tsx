import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import { PatientInfoLonghand } from './PatientInfoLonghand';

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: 'Patient Information (longhand)',
  component: PatientInfoLonghand,
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {
    options: { 
      name:"Options",
      type:{name:"array", required:true}
    },
  },
} as ComponentMeta<typeof PatientInfoLonghand>;

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof PatientInfoLonghand> = (args) => <PatientInfoLonghand {...args} />;

export const Standard = Template.bind({});
// More on args: https://storybook.js.org/docs/react/writing-stories/args
Standard.args = {
    hospitalIdentifier: "MRN0000000",
    name: "first last",
    dateOfBirth: "01/01/1970"
};