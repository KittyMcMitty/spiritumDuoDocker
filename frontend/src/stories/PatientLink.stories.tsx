import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import PatientLink  from './PatientLink';
import StoryRouter from 'storybook-react-router';


// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: 'Patient Link',
  component: PatientLink,
  decorators: [ StoryRouter() ],
} as ComponentMeta<typeof PatientLink>;

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof PatientLink> = (args) => <PatientLink {...args} />;

export const Default = Template.bind({});
// More on args: https://storybook.js.org/docs/react/writing-stories/args
Default.args = {
    patientId: "MRN1234567",
    name: "John Doe"
};
