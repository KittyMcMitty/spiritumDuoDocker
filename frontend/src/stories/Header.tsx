import React from 'react';

import { Button } from './Button';
import './header.css';
import { PathwaySelector } from './PathwaySelector';

interface HeaderProps {
	user?: {};
}

export const Header = ({user}: HeaderProps) => (
	<nav className="navbar navbar-expand-lg navbar-light bg-light">
	<div className="container-fluid">
		<form className="d-flex">
			<input className="form-control me-2" type="search" placeholder="Hospital number" aria-label="Hospital number"/>
		</form>
		<div className="collapse navbar-collapse" id="navbarSupportedContent">
			<ul className="navbar-nav mb-2">
				<li className="nav-item">
					<a className="nav-link active" aria-current="page" href="#">Home</a>
				</li>
				<li className="nav-item">
					<a className="nav-link" href="#">Triage</a>
				</li>
				<li className="nav-item">
					<a className="nav-link" href="#">Clinic</a>
				</li>
				<li className="nav-item">
					<a className="nav-link" href="#">MDT</a>
				</li>
				<li className="nav-item">
					<a className="nav-link" href="#">Refer</a>
				</li>
			</ul>
		</div>
		<PathwaySelector options={["Lung cancer", "Bronchieactasis"]} onItemSelect={(data:string)=>console.log(data)}/>
	</div>
</nav>
);