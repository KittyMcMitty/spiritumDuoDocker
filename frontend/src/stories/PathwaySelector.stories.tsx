import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import { PathwaySelector } from './PathwaySelector';

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: 'Pathway Selector',
  component: PathwaySelector,
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {
    options: { 
      name:"Options",
      type:{name:"array", required:true}
    },
  },
} as ComponentMeta<typeof PathwaySelector>;

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof PathwaySelector> = (args) => <PathwaySelector {...args} />;

export const Standard = Template.bind({});
// More on args: https://storybook.js.org/docs/react/writing-stories/args
Standard.args = {
  options: ["Lung cancer", "Bronchieactasis"],
};