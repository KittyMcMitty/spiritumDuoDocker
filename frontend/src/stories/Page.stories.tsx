import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import { Page } from './Page';
import * as HeaderStories from './Header.stories';

export default {
  title: 'Example/Page',
  component: Page,
} as ComponentMeta<typeof Page>;

const Template: ComponentStory<typeof Page> = (args) => <Page {...args} />;

export const Standard = Template.bind({});