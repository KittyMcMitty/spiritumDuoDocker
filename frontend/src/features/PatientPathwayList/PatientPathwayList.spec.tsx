import React from 'react';
import { composeStories } from '@storybook/testing-react';
import { render, waitFor } from '@testing-library/react';
import * as stories from './PatientPathwayList.stories';

const { Default } = composeStories(stories);

it('It should render Default correctly', () => {
  const { container } = render(<Default />);
  expect(container).toMatchSnapshot();
});

it('Should have accessible table', async () => {
  const { getByRole } = render(<Default />);
  await waitFor(() => expect(getByRole('table', { name: /pathway\s+table/i })).toBeInTheDocument());
});
