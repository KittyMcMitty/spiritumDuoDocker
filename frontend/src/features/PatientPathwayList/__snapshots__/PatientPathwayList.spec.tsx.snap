// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should render Default correctly 1`] = `
<div>
  <div
    style="width: 90vw; height: 100vh; min-width: 300px;"
  >
    <div>
      Loading
    </div>
  </div>
</div>
`;
